File: Go/leetcode/1.go
package leetcode

//make(type,len,capacity=len)
//访问数组时不得越界(len)
//切片不得越界(capacity)
//切片共用地址空间

//make(map[type1]type2)对应的map有默认值
//map[type1]type2
//m[type1]=type2
import(
	"strconv"
)
func twoSum(nums []int, target int) []int {
	m := make(map[int]int)
	for i := 0;i < len(nums);i++{
		another := target - nums[i]
		if _,ok := m[another];ok{
			return []int{m[another],i}
		}
		m[nums[i]]= i
	}
	return nil
}

func TwoSum(nums []int, target int) []int {
	return twoSum(nums,target)
}

type ListNode struct {
    Val int
	Next *ListNode
}
func AddTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{Val: 0}
	n1,n2,carry,current := 0,0,0,head
	for l1 != nil || l2 != nil || carry != 0{
		if l1 != nil{
			 n1=l1.Val
			 l1=l1.Next
		 }else{
			 n1=0
		 }
		 if l2 != nil{
			 n2=l2.Val
			 l2=l2.Next
		 }else{
			 n2=0
		 }
		 current.Next = &ListNode{Val: (n1+n2+carry)%10}
		 current = current.Next
		 carry = (n1+n2+carry)/10
	}
	return head.Next
}
func lengthOfLongestSubstring(s string)int{
	if len(s) == 0{
		return 0
	}
	var bitSet [256]bool
	result,left,right := 0,0,0
	for right < len (s){
		if bitSet[s[right]] {
			bitSet[s[left]]=false
			left++
		}else{
			bitSet[s[right]]=true
			right++
		}
		if result < right-left{
			result=right-left
		}
		if left + result >= len(s){
			break
		}
	}
	return result
}
func FindMedianSortedArrays(nums1 []int,nums2 []int)float64{
	//保证len(nums1)<=len(nums2)
	if len(nums1) > len(nums2){
		return FindMedianSortedArrays(nums2,nums1);
	}
	low,high,k,nums1Mid,nums2Mid := 0,len(nums1),(len(nums1)+len(nums2)+1)>>1,0,0
	for low <= high{
		nums1Mid = low + (high-low)>>1
		nums2Mid = k - nums1Mid
		if nums1Mid > 0 && nums1[nums1Mid-1] > nums2[nums2Mid]{
			high = nums1Mid - 1;
		}else if nums1Mid != len(nums1) && nums1[nums1Mid] < nums2[nums2Mid - 1]{
			low = nums1Mid + 1;
		}else{
			break
		}
	}
	midLeft,midRight := 0,0
	if nums1Mid == 0{
		midLeft = nums2[nums2Mid - 1]
	}else if nums2Mid == 0{
		midLeft = nums1[nums1Mid - 1]
	}else{
		midLeft = max(nums1[nums1Mid - 1],nums2[nums2Mid - 1])
	}
	if (len(nums1)+len(nums2))&1 == 1{
		return float64(midLeft)
	}
	if nums1Mid == len(nums1){
		midRight = nums2[nums2Mid]
	}else if nums2Mid == len(nums2){
		midRight = nums1[nums1Mid]
	}else{
		midRight = nums1[nums1Mid] ^ nums2[nums2Mid] ^ max(nums1[nums1Mid],nums2[nums2Mid])
	}
	return float64(midLeft+midRight)/2
}

func max(a,b int)int{
    if a > b{
        return a
    }
    return b
}

func min(a,b int)int{
	if a < b{
		return a
	}
	return b
}

func longestPalindrome1(s string) string {
    if len(s) < 2{
		return s
	}
	newS := make([]rune,0)
	newS = append(newS,'#')
	for _,c := range s{
		newS = append(newS,c)
		newS = append(newS,'#')
	}
	dp,maxRight,center,maxLen,begin := make([]int,len(newS)),0,0,1,0
	for i := 0;i < len(newS);i++{
		if i < maxRight{
			dp[i] = min(dp[2*center-i],maxRight-i)
		}
		left,right := i - (1+dp[i]),i + (1+dp[i])
		for left >= 0 && right < len(newS) && newS[left] == newS[right]{
			dp[i]++
			left--
			right++
		}
		//update the maxRight
		if dp[i] + i > maxRight{
			maxRight = dp[i] + i
			center = i
		}
		//update the maxLen
		if dp[i] > maxLen{
			maxLen=dp[i]
			begin = (i - maxLen)/2
		}
	}
	return s[begin : begin + maxLen]
}

//滑动窗口解法
func longestPalindrome2(s string) string {
	left,right,pl,pr := 0,0,0,0
	for left < len(s){
		for right + 1 < len(s) && s[left] == s[right+1]{
			right++
		}
		for left > 0 && right < len(s) - 1 && s[left-1] == s[right+1]{
			left--;
			right++;
		}
		if right -left > pr - pl{
			pr = right
			pl = left
		}
        left = (left + right)/2 + 1
        right = left
	}
	return s[pl : pr + 1]
}

//中心扩散法
func longestPalindrome(s string) string {
	if len(s) < 2{
		return s
	}
	res := ""
	for i := 0;i < len(s) - 1;i++{
		res = maxPalindrome(s,i,i,res)
        if s[i] == s[i+1]{
			res = maxPalindrome(s,i,i+1,res)
		}
	}
	return res
}
func maxPalindrome(s string,i,j int,res string)string{
	sub := ""
	for i >= 0 && j < len(s) && s[i] == s[j]{
		i--
		j++
	}
    sub = s[i+1 : j]
    if len(sub) > len(res){
		return sub
	}
	return res
}

func convert(s string, numRows int) string {
    matrix,down,up := make([][]byte,numRows,numRows),0,numRows-2
	for i := 0;i != len(s);i++{
		if down != numRows{
			matrix[down] = append(matrix[down],byte(s[i]))
			down++
		}else if up > 0{
			matrix[up] = append(matrix[up],byte(s[i]))
			up--
		}else{
			down = 0
			up =numRows-2
			i--
		}
	}
	res := make([]byte,0,len(s))
	for _,row := range matrix{
		for _,item := range row{
			res =  append(res,item)
		}
	}
    return string(res)
}

func reverse(x int) int {
	res := 0
	for x != 0{
		res = res*10 + x%10
		x /= 10
	}
	//1<<31等于2^31(针对于Go)
    if res > 1<<31-1 || res < -(1<<31){
		return 0
	}
	return res
}

func MyAtoi(s string) int {
    allowSpace,allowSign,maxInt,sign,digits := true,true,int64(1<<31),1,[]int{}
	for _,c := range s{
		if c == ' ' && allowSpace{
			continue
		}
		if allowSign{
			allowSign,allowSpace = false,false
			if c == '-'{
				sign = -1
				continue
			}
			if c == '+'{
				continue
			}
		}
		if c < '0' || c > '9'{
			break
		}
		digits = append(digits,int(c-48))
	}
	var num,place int64
	place,num = 1,0
	lastLeadingZeroIndex := -1
	for i,d := range digits{
		if d == 0 {
			lastLeadingZeroIndex = i
		}else{
			break
		}
	}
	if lastLeadingZeroIndex > -1{
		digits = digits[lastLeadingZeroIndex+1 : ]
	}
	var rtnMax int64
    if sign > 0{
        rtnMax = maxInt - 1
    }else{
        rtnMax = maxInt
    }
	digitsCount := len(digits)
	for i := digitsCount -1;i >= 0;i--{
		num += int64(digits[i])*place
		place *= 10
		if digitsCount -i > 10 || num > rtnMax{
			return int(int64(sign) * rtnMax)
		}
	}
	return int(num*int64(sign))
}

func isPalindrome(x int)bool{
	if x < 0{
		return false
	}
	if x == 0{
		return true
	}
	if x%10 == 0{
		return false
	}
	arr := make([]int,0,12)
	for x > 0{
		arr = append(arr,x%10)
		x /= 10
	}
	sz := len(arr)
	for i,j := 0,sz-1;i < j;i,j = i+1,j-1{
		if arr[i] != arr[j]{
			return false
		}
	}
	return true
}

func isPalindrome1(x int)bool{
	if x < 0{
		return false
	}
	if x < 10{
		return true
	}
	if x%10 == 0{
		return false
	}
	s := strconv.Itoa(x)
	length := len(s)
	for i := 0;i < length/2;i++{
		if s[i] != s[length-i-1]{
			return false
		}
	}
	return true
}

File: c++/1.cpp
#include<iostream>
#include<vector>
#include<map>
#include<algorithm>
#include<cstring>
#include<regex>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> ans;
        map<int,int> m;
        for(int i=0;i<nums.size();i++){
            int another=target-nums[i];
            if(m[another]){
                ans.push_back(m[another]-1);
                ans.push_back(i);
                break;
            }
            m[nums[i]]=i+1;
        }
        return ans;
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode();
        int n1=0,n2=0,carry=0;
        ListNode *current=head;
        while(l1!=nullptr || l2 != nullptr || carry != 0){
            if(l1==nullptr)n1=0;
            else{
                n1=l1->val;
                l1=l1->next;
            }
            if(l2==nullptr)n2=0;
            else{
                n2=l2->val;
                l2=l2->next;
            }
            current->next=new ListNode((n1+n2+carry)%10);
            current=current->next;
            carry=(n1+n2+carry)/10;
        }
        return head->next;
    }
    int lengthOfLongestSubstring(string s) {
        //使用map
        map<int,int> m;
        //cur_sta表示当前子串起点
        int ans=0,cur_sta=0,len=s.size();
        for(int i=0;i<len;i++){
            if(m[(int)s[i]]){
                //如果当前字符已经出现过
                ans=max(i-cur_sta,ans);
                //擦除
                for(int j=cur_sta;j<m[(int)s[i]]-1;j++){
                    m[(int)s[j]]=0;
                }
                cur_sta=m[(int)s[i]];
            }
            m[(int)s[i]]=i+1;
        }
        return max(ans,len-cur_sta);
    }
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int len1=nums1.size(),len2=nums2.size();
        vector<int> ans;
        int cur1=0,cur2=0;
        while(cur1 < len1 || cur2 < len2){
            while(cur1< len1 && (cur2 ==len2 || nums1[cur1] <= nums2[cur2])){
                ans.push_back(nums1[cur1]);
                cur1++;
            }
            while(cur2 < len2 && (cur1 == len1 || nums1[cur1] > nums2[cur2])){
                ans.push_back(nums2[cur2]);
                cur2++;
            }
        }
        cout<<ans.size()<<endl;
        for(int i=0;i<ans.size();i++){
            cout<<ans[i]<<endl;
        }
        if((ans.size())%2==0)return (ans[ans.size()/2-1]+ans[ans.size()/2])/2.0;
        else return ans[ans.size()/2];
    }
    //dp
    string longestPalindrome1(string s) {
        const int N=1005;
        bool a[N][N]={0};
        int sta=0,ans=1;
        for(int i=0;i<s.size();i++){
            for(int j=0;j<s.size()-i;j++){
                if(s[j]==s[j+i] && (j+1 >= j+i-1 || a[j+1][j+i-1])){
                    a[j][j+i]=true;
                    if(i+1>ans){
                        ans=i+1;
                        sta=j;
                    }
                }
            }
        }
        return s.substr(sta,ans);
    }

//马拉车算法
    string longestPalindrome(string s){
        if(s.size() < 2)return s;
        const int N=2005;
        int a[N]={0};
        char cc[N];
        int center=0,maxRight=0,maxLen=1,sta=0;
        //预处理
        cc[0]='#';
        for(int i=0;i<s.size();i++){
            cc[i*2+1]=s[i];
            cc[2*(i+1)]='#';
        }
        for(int i=0;i < s.size()*2+1;i++){
            if(i < maxRight){
                //核心算法
                a[i] = min(a[2*center-i],maxRight-i);
            }
            int left=i-(a[i]+1),right=i+(a[i]+1);
            while(left >= 0 && right < s.size()*2+1 && cc[left] == cc[right]){
                left--;
                right++;
                a[i]++;
            }
            //更新maxRight
            if(a[i] + i > maxRight){
                maxRight =a[i] + i;
                center = i;
            }
            //更新答案
            if(a[i] > maxLen){
                maxLen = a[i];
                sta = (i - a[i])/2;
            }
        }
        return s.substr(sta,maxLen);
    }

//题意
// 将给定的字符串按如下的Z字型排列，并从左至右输出
// |  /  /
// | /| /|
// |/ |/
// |  |
    string convert(string s, int numRows) {
        vector<vector<char>> ans(numRows);
        int down=0,up=numRows-2;
        for(int i=0;i < s.size();i++){
            if(down < numRows){
                ans[down].push_back(s[i]);
                down++;
            }else if(up > 0){
                ans[up].push_back(s[i]);
                up--;
            }else{
                down=0;
                up=numRows-2;
                i--;
            }
        }
        string solution="";
        for(int i=0;i<numRows;i++){
            for(int j=0;j < ans[i].size();j++){
                solution.push_back(ans[i][j]);
            }
        }
        return solution;
    }

//注意最大值和最小值
    int reverse(int x) {
        long long ans=0;
        int maxx=-1*((1<<31)+1),minn=(1<<31);
        while(x != 0){
            ans=ans*10+x%10;
            x/=10;
        }
        if(ans > maxx || ans < minn)return 0;
        return ans;
    }

    int myAtoi(string s) {
        int maxx=-1*((1<<31)+1),minn=1<<31;
        bool allowSign=true,allowSpace=true;
        int sign=1;
        vector<int> ans;
        for(int i=0;i<s.size();i++){
            if(s[i]==' '){
                if(allowSpace)continue;
                else break;
            }else if((s[i] == '+' || s[i] == '-')){
                if(!allowSign)break;
                if(s[i] == '-')sign = -1;
                allowSign=false;
                allowSpace=false;
            }else if(s[i] <= '9' && s[i] >= '0'){
                allowSign=false;
                allowSpace=false;
                ans.push_back((int)(s[i]-'0'));
            }else{
                break;
            }
        }
        long long res=0;
        bool leadingZero=true;
        int substitute = sign > 0 ? maxx : minn;
        for(int i=0;i<ans.size();i++){
            if(!ans[i] && leadingZero)continue;
            if(ans.size() - i > 10)return substitute;
            leadingZero=false;
            res=res*10+ans[i];
        }
        if(res > maxx || res < minn)return substitute;    
        return sign*res;
    }

    bool isPalindrome(int x) {
        if(x < 0)return false;
        vector<int> res;
        while(x){
            res.push_back(x%10);
            x/=10;
        } 
        int len=res.size();
        for(int i=0;i<len/2;i++){
            if(res[i] != res[len-i-1])
                return false;
        }    
        return true;
    }
    //通过分支判断无法实现功能
    // bool isMatch(string s, string p) {
    //     int ps=0,pp=0;
    //     //对p中*前后的字符进行预处理
    //     for(int i=2;i<p.size();){
    //         if(p[i-1] == '*' && (p[i] == '*' || p[i] == p[i-2]))p.erase(p.begin()+i);
    //         else i++;
    //     }
    //     int lens=s.size(),lenp=p.size();
    //     while(ps < lens && pp < lenp){
    //         //直接出现*视为无效
    //         if(p[pp] == '*'){
    //             pp++;
    //             continue;
    //         }
    //         if(pp+1 < lenp && p[pp+1] == '*'){
    //             while(ps < lens && (s[ps] == p[pp] || p[pp] == '.'))ps++;
    //             pp+=2;
    //         }else{
    //             if(s[ps] != p[pp] && p[pp] != '.')return false;
    //             ps++;
    //             pp++;
    //         }
    //     }
    //     if(ps < lens || pp < lenp)return false;
    //     return true;
    // }
    bool isMatch1(string s, string p) {
        return regex_match(s,regex(p));
    }
    // dp[a][b]表示将s[a]与p[b]相匹配，若返回false，表示不能将s[a]与p[b]相匹配
    bool solver(string &s,string &p,int a ,int b,vector<vector<int>> &dp)  {   
        // 递归终止条件
        //若dp[a][b]!=-1,表示已经计算过,return
        if(dp[a][b]!=-1) return dp[a][b];
        //匹配完毕,return true
        if(a>=s.length() && b>=p.length())  return dp[a][b] = true;
        //p已经使用完但s还有未匹配的,return false
        if(b>=p.length())  return dp[a][b] = false;
        
        //s[a]能与p[b]相匹配
        bool match = ( a <s.length() && (s[a]==p[b] || p[b]=='.'));
        //关键算法，若出现*通配符，可选择不匹配或选择匹配
        if( b<p.length()-1 && p[b+1]=='*') return dp[a][b] = solver(s,p,a,b+2,dp) || ( match && solver(s,p,a+1,b,dp));
        //普通匹配
        if(match) return dp[a][b] = solver(s,p,a+1,b+1,dp);
        return dp[a][b] = false;
    }
    bool isMatch(string s, string p) {
        vector<vector<int>> dp( s.length()+1, vector<int>(p.length()+1,-1) );
        return solver(s,p,0,0,dp);
    }
};

int main(){
    vector<int> ans(2);//创建一个size=2的vector
    string s="aacabdkacaa";
    map<int,int> m;
    m[1]=2;
    // cout<<m[1];
    m.erase(1);
    // cout<<m[1];
    // cout<<(int)s[1];
    Solution solution=Solution();
    // cout<<solution.lengthOfLongestSubstring(s);
    vector<int> a,b;
    a.push_back(1),a.push_back(3);
    b.push_back(2);
    // solution.findMedianSortedArrays(a,b);
    // cout<<solution.longestPalindrome1(s);
    // s.append("#");
    // cout<<solution.longestPalindrome(s)<<endl;
    string ss="PAYPALISHIRING";
    // cout<<solution.convert(ss,3);
    // cout<<solution.reverse(-123);
    // cout<<solution.myAtoi("-0042");
    // cout<<solution.isPalindrome(1001);
    cout<<solution.isMatch("aa","a*");
    return 0;
}